File: src/main/java/com/csl/thread/ThreadTester.java
package com.csl.thread;

public class ThreadTester {
    public static void main(String[] args) {

        for (int i=0; i<10; i++) {
            System.out.println("main: "+i);
        }

        //使用暱名類別
        Thread thread = new Thread() {
            @Override
            public void run() {
                for (int i=0; i<10 ; i++) {
                    System.out.println("com/csl/thread " + i);
                    try {
                        sleep(50);
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                }
            }
        };
        thread.start();

        MyThread myThread = new MyThread();
        myThread.start();

        //使用介面Runnable
        Runnable runnable = new Runnable() {
            @Override
            public void run() {
                for (int i=0; i<10; i++) {
                    System.out.println("Runnable "+i);

                    try {
                        Thread.sleep(50);
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                }
            }
        };
        Thread r = new Thread(runnable);
        r.start();

        //使用lambda
         new Thread(()-> {
            for (int i=0; i<10; i++) {
                System.out.println("com.csl.thread runnable lambda " + i);

                try {
                    Thread.sleep(50);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }
        }).start();

        System.out.println("main end");
    }
}

class MyThread extends Thread {
    @Override
    public void run() {
        for (int i=0; i<10 ; i++) {
            System.out.println(getName() + " " + i);

            try {
                sleep(50);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    }
}

File: src/main/kotlin/com/csl/thread/ThreadTester.kt
package com.csl.thread.kt

import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlin.concurrent.thread

fun main(args:Array<String>) {

    //輕量化thread 更快速 -> coroutines
    // implementation 'org.jetbrains.kotlinx:kotlinx-coroutines-core:1.2.0-alpha-2'
    GlobalScope.launch {
        for (i in 1..100)
            println("Coroutines: ${i}")

        delay(50)
    }


    //way 1
    val runnable:Runnable = Runnable {
        for (i in 1..100)
            println("Runnable: "+ i)
        Thread.sleep(50)
    }
    Thread(runnable).start()


    //way 2
    Thread({
        for (i in 1..100)
            println("Thread: "+ i)
        Thread.sleep(50)
    }).start()

    //way 3
    thread {
        for (i in 1..100)
            println("com.csl.thread(): "+ i)
        Thread.sleep(50)
    }




}
